File: source/materials/metal.cpp
#include "materials/metal.hpp"

metal::metal(colour const &colour_in, double roughness_in) noexcept : super(colour_in), roughness(roughness_in) {}

auto metal::scatter(const ray &ray_in, const hit &hit_in) const -> std::optional<scatter_data>
{
   vec reflected = reflect(ray_in.direction() / ray_in.direction().length(), hit_in.normal);
   ray scattered = ray(hit_in.position, reflected + roughness * random_in_unit_sphere());

   if (dot(reflected, hit_in.normal) > 0)
   {
      return std::make_pair(super::emission, scattered);
   }
   else
   {
      return {};
   }
}

File: go-rt/materials/metal.go
package materials

import (
	"go_rt/core"
	"go_rt/maths"
)

type Metal struct {
    Diffuse core.Colour

    Roughness float64
}

func (this Metal) Scatter(ray *core.Ray, hit *core.RayHit) ScatterData {
    unitDir := maths.Vec3Normalise(&ray.Direction)
    reflected := maths.Reflect(&hit.Normal, &unitDir)  

    random := maths.RandomVec3InUnitSphere()
    random.MultScalar(this.Roughness)

    scattered := core.Ray{Origin: hit.Position, Direction: maths.Add(&reflected, &random)}

    if maths.Vec3Dot(&scattered.Direction, &hit.Normal) > 0 {
        return ScatterData{
            Emission: core.Colour{},
            Diffuse: this.Diffuse,
            Ray: scattered}
    } else {
        return ScatterData{
            Emission: core.Colour{},
            Diffuse: core.Colour{},
            Ray: scattered}
    }
}
